test/csuite/wt8963_insert_stress/main.c

```diff
@@ -0,0 +1,171 @@
+ * software to the public domain. We make this dedication for the benefit
+ * of the public at large and to the detriment of our heirs and
+ * successors. We intend this dedication to be an overt act of
+ * relinquishment in perpetuity of all present and future rights to this
+ * software under copyright law.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+#include "test_util.h"
+
+/*
+ * The motivation for this test is to try and reproduce BF-24385 by stressing insert functionality.
+ * The test creates a lot of threads that concurrently insert a lot of records with random keys.
+ * Having a large memory page ensures that we have big insert lists. Big cache size allows having
+ * more dirty content in the memory before eviction kicks in.
+ */
+
+#define THREAD_NUM_ITERATIONS 200000
+#define NUM_THREADS 110
+#define KEY_MAX UINT32_MAX
+
+void *thread_insert_race(void *);
+
+static uint64_t ready_counter;
+static WT_RAND_STATE rnd;
+
+/*
+ * set_key --
+ *     Wrapper providing the correct typing for the WT_CURSOR::set_key variadic argument.
+ */
+static void
+set_key(WT_CURSOR *c, uint64_t value)
+{
+    c->set_key(c, value);
+}
+
+/*
+ * set_value --
+ *     Wrapper providing the correct typing for the WT_CURSOR::set_value variadic argument.
+ */
+static void
+set_value(TEST_OPTS *opts, WT_CURSOR *c, uint64_t value)
+{
+    if (opts->table_type == TABLE_FIX)
+        c->set_value(c, (uint8_t)value);
+    else
+        c->set_value(c, value);
+}
+
+/*
+ * main --
+ *     Test's entry point.
+ */
+int
+main(int argc, char *argv[])
+{
+    TEST_OPTS *opts, _opts;
+    WT_CURSOR *cursor;
+    WT_SESSION *session;
+    clock_t ce, cs;
+    pthread_t id[NUM_THREADS];
+    int i, ret;
+    char tableconf[128];
+
+    opts = &_opts;
+    memset(opts, 0, sizeof(*opts));
+    opts->nthreads = NUM_THREADS;
+    opts->table_type = TABLE_ROW;
+    testutil_check(testutil_parse_opts(argc, argv, opts));
+    testutil_make_work_dir(opts->home);
+
+    testutil_check(
+      wiredtiger_open(opts->home, NULL, "create,cache_size=3G,statistics=(fast)", &opts->conn));
+    testutil_check(opts->conn->open_session(opts->conn, NULL, NULL, &session));
+    testutil_check(__wt_snprintf(tableconf, sizeof(tableconf),
+      "key_format=%s,value_format=%s,leaf_page_max=32k,memory_page_image_max=50MB",
+      opts->table_type == TABLE_ROW ? "Q" : "r", opts->table_type == TABLE_FIX ? "8t" : "Q"));
+    testutil_check(session->create(session, opts->uri, tableconf));
+
+    __wt_random_init_seed((WT_SESSION_IMPL *)session, &rnd);
```

Comment:
Should we give every thread its own random number generator? (What format does is alternate between a single RNG and a per-thread RNG, and we've anecdotally seen that find different errors.)

Reply:
I originally did that but then I thought that it would be safer to seed it once and use everywhere to there's no 100% overlap in random numbers in threads. But now that I looked at the implementation of `__wt_random_init_seed` it seems quite impossible to have the same WT_RAND_STATE::v. Changed it: 2fecb8bc2a1dd73d77bc4207b14ed0bcb09f6168

Reply:
I was thinking more about the cache shootdowns caused by using the same see in lots of threads.

Reply:
I see, thank you for the explanation.